test-runner-frontend/src/Views/Dashboard.js

```diff
@@ -6,27 +6,35 @@ import Stack from '@mui/material/Stack';
 import GraphQL_Logo from '../Assets/GraphQL_Java_Logo_v2.png';
 import Alert from '@mui/material/Alert';
 import { useEffect, useState } from 'react';
-//import { onSnapshot, collection } from '@firebase/firestore';
-//import db from './firebase';
+import { onSnapshot, collection } from '@firebase/firestore';
+import db from './firebase';
 import TestRunsTable from '../Components/TestRunsTable';
 //import { useNavigate } from 'react-router-dom';
 import InputLabel from '@mui/material/InputLabel';
 import MenuItem from '@mui/material/MenuItem';
 import FormControl from '@mui/material/FormControl';
 import Select from '@mui/material/Select';
+import CircularProgress from '@mui/material/CircularProgress';
+import {
+  getMachineNames,
+  sortTestRunsByMachine,
+  getBenchmarksByMachine,
+  flattenSortedTestRuns,
+  getAllBenchmarks,
+} from './DashboardUtils';
 
 export default function Dashboard() {
   const [isCheckBoxActive, setCheckboxActiveState] = React.useState(false);
-
   //const [cancelButtonState, setCancelButtonState] = React.useState(false);
+  const [testResults, setTestResults] = useState([]);
+  const [loadingState, setLoadingState] = useState(true);
+  const [testRunResults, setTestRunResults] = useState([]);
+  const [testRunResultsCopy, setTestRunResultsCopy] = useState([]);
+  const [testRunSelection, setTestRunSelection] = useState('All Test Runs');
+  const [machineSelection, setMachineSelection] = useState('All Machines');
+  const [checkBoxSelection, setCheckBoxSelection] = React.useState([]);
```

Comment:
we now have a lot of different states. it would be worth looking into using a reducer (https://reactjs.org/docs/hooks-reference.html#usereducer) so its only one blob of state

Reply:
Done.